src/quick-lint-js/parse.h

```diff
@@ -2495,6 +2512,45 @@ class parser {
     }
   }
 
+  std::optional<function_attributes> try_parse_function_with_leading_star() {
+    QLJS_ASSERT(this->peek().type == token_type::star);
+    token star_token = this->peek();
+    lexer_transaction transaction = this->lexer_.begin_transaction();
+    this->skip();
+    if (!this->peek().has_leading_newline) {
+      function_attributes attrb = function_attributes::generator;
+      bool has_leading_async = this->peek().type == token_type::kw_async;
+      // *async function f() {}
+      if (has_leading_async) {
+        attrb = function_attributes::async_generator;
+        this->skip();
+      }
+      // *function f() {}
+      if (this->peek().type == token_type::kw_function) {
+        this->skip();
+        if (this->peek().type == token_type::identifier) {
+          this->error_reporter_->report(
+              error_generator_function_star_belongs_before_name{
+                  .function_name = this->peek().span(),
+                  .star = star_token.span(),
+              });
+        } else {
+          this->error_reporter_->report(
+              error_generator_function_star_belongs_after_keyword_function{
+                  .star = star_token.span()});
+        }
+        this->lexer_.roll_back_transaction(std::move(transaction));
+        this->skip();
+        if (has_leading_async) {
+          this->skip();
+        }
+        return attrb;
+      }
+    }
+    this->lexer_.roll_back_transaction(std::move(transaction));
+    return std::nullopt;
```

Comment:
Instead of nesting `if`s so much, let's use early returns. Example (I hope this diff is readable...):
```suggestion
    if (this->peek().has_leading_newline) {
      this->lexer_.roll_back_transaction(std::move(transaction));
      return std::nullopt;
    }

      function_attributes attrb = function_attributes::generator;
      bool has_leading_async = this->peek().type == token_type::kw_async;
      // *async function f() {}
      if (has_leading_async) {
        attrb = function_attributes::async_generator;
        this->skip();
      }
      // *function f() {}
      if (this->peek().type != token_type::kw_function) {
        this->lexer_.roll_back_transaction(std::move(transaction));
        return std::nullopt;
      }

        this->skip();
        if (this->peek().type == token_type::identifier) {
          this->error_reporter_->report(
              error_generator_function_star_belongs_before_name{
                  .function_name = this->peek().span(),
                  .star = star_token.span(),
              });
        } else {
          this->error_reporter_->report(
              error_generator_function_star_belongs_after_keyword_function{
                  .star = star_token.span()});
        }
        this->lexer_.roll_back_transaction(std::move(transaction));
        this->skip();
        if (has_leading_async) {
          this->skip();
        }
        return attrb;
```